students/DrewSmith/lessons/lesson09/assignment/charges_calc.py

```diff
@@ -0,0 +1,185 @@
+            print(f"WARNING: {key} missing {key_err.args[0]} value")
+        except ValueError as value_error:
+            logging.warn(f"Invalid data: {key}: {str(value_error)}")
+            print(f"WARNING: {key} has invalid price_per_day")
+
+        try:
+            value['sqrt_total_price'] = math.sqrt(value['total_price'])
+        except ValueError as value_error:
+            logging.warn(f"Invalid data: {key}: {str(value_error)}")
+            print(
+                f"WARNING: {key} unable square root the negative total_price")           
+        
+        try:
+            value['unit_cost'] = value['total_price'] / value['units_rented']
+        except KeyError as key_err:
+            logging.warn(key_log.format(key, key_err.args[0], str(key_err)))
+            print(f"WARNING: {key} missing {key_err.args[0]} value")
+        except ValueError as value_error:
+            logging.warn(f"Invalid data: {key}: {str(value_error)}")
+            print(f"WARNING: {key} has invalid units_rented value")
+        except ZeroDivisionError as zero_error:
+            logging.warn(
+                f"Invalid data: {key}: units_rented is 0: {str(zero_error)}")
+            print(f"WARNING: {key} units_rented cannot be 0")
+
+def save_to_json(filename, file_data):
+    '''
+    Outputs the results to the JSON ouput file
+
+    :param filename: output file name
+    :param data: data to write to file
+    '''
+    with open(filename, 'w') as file:
+        json.dump(file_data, file, indent=4)
+        logging.debug(f"Output file generated: {filename}")
+        print("Output file complete")
+
+def parse_cmd_arguments():
+    ''' Processes commandline arguments from the user '''
+    parser = argparse.ArgumentParser(description='Process some integers.')
+    parser.add_argument('-i', '--input', help='input JSON file', required=True)
+    parser.add_argument('-o', '--output', help='ouput JSON file', required=True)
+    parser.add_argument('-d', '--debug', help='ouput JSON file', required=False)
+
+    parser.add_argument('-l', '--log_off', help='ouput JSON file', required=False)
+
+    result_args = parser.parse_args()
+    load_logging(result_args.debug)
+
+    logging.debug(f"Input: {result_args.input}")
+    logging.debug(f"Output: {result_args.output}")
+    logging.debug(f"Debug: {result_args.debug}")
+    logging.debug(f"log_off: {result_args.log_off}")
+
+    return result_args
+
+if __name__ == "__main__":
+    ARGS = parse_cmd_arguments()
+
+    # Selective method logging deactivate
+    if ARGS.log_off is not None:
+        if 'load' in ARGS.log_off.lower():
+            load_rentals_file = disable_method_logging(load_rentals_file)
+        if 'calc' in ARGS.log_off.lower():
+            calculate_additional_fields = disable_method_logging(calculate_additional_fields)
+        if 'save' in ARGS.log_off.lower():
+            save_to_json = disable_method_logging(save_to_json)
```

Comment:
Interesting approach. It works well since this file is all about using the command line, and this is doing more than the assignment asked for (I think they were assuming that the switch in the args would be more like an on/off, and would only affect the decorated functions, but would not decide which ones were decorated). But, this is an appropriate solution, given the criteria. Note that if you actually wanted to be able to import these functions in to another file, it would be better to use the syntactic sugar that is the @ symbol, but then you would have a whole new set of problems regarding who is controlling those decorators. :)